chemprop/nn/agg.py

```diff
@@ -112,6 +120,10 @@ def __init__(self, dim: int = 0, *args, output_size: int, **kwargs):
         self.W = nn.Linear(output_size, 1)
 
     def forward(self, H: Tensor, batch: Tensor) -> Tensor:
-        alphas = scatter_softmax(self.W(H), batch, self.dim)
-
-        return scatter(alphas * H, batch, self.dim, reduce="sum")
+        index_torch = batch.unsqueeze(1).repeat(1, H.shape[1])
+        dim_size = batch.max().int() + 1
+        H_exp = H.exp()
+        Z = torch.zeros(dim_size, H.shape[1], dtype=H.dtype, device=H.device).scatter_reduce_(
+            self.dim, index_torch, H_exp, reduce="sum", include_self=False
+        )
+        return H_exp / Z[batch]
```

Comment:
This isn't quite correct. Note how it doesn't make use of `self.W` while the current v2/dev version does. 
You can see a dimension mismatch error by replacing `agg = nn.MeanAggregation()` in `examples/training.ipynb` with `agg = nn.AttentiveAggregation(output_size=mp.output_dim)`.

Here is a version of forward that works by my tests.

```
def forward(self, H: Tensor, batch: Tensor) -> Tensor:
    dim_size = batch.max().int() + 1
    attention_logits = (self.W(H) - self.W(H).max()).exp()
    Z = torch.zeros(dim_size, 1, dtype=H.dtype, device=H.device).scatter_reduce_(
        self.dim, batch.unsqueeze(1), attention_logits, reduce="sum", include_self=False
    )
    alphas = attention_logits / Z[batch]
    index_torch = batch.unsqueeze(1).repeat(1, H.shape[1])
    return torch.zeros(dim_size, H.shape[1], dtype=H.dtype, device=H.device).scatter_reduce_(
        self.dim, index_torch, alphas * H, reduce="sum", include_self=False
    )
```

I tested it by replacing `agg = nn.AttentiveAggregation(output_size=mp.output_dim)` in `examples/training.ipynb` with:
```
import torch
from torch import Tensor
class NewAgg(nn.AttentiveAggregation):
    def forward(self, H: Tensor, batch: Tensor) -> Tensor:
        dim_size = batch.max().int() + 1
        attention_logits = (self.W(H) - self.W(H).max()).exp()
        Z = torch.zeros(dim_size, 1, dtype=H.dtype, device=H.device).scatter_reduce_(
            self.dim, batch.unsqueeze(1), attention_logits, reduce="sum", include_self=False
        )
        alphas = attention_logits / Z[batch]
        index_torch = batch.unsqueeze(1).repeat(1, H.shape[1])
        return torch.zeros(dim_size, H.shape[1], dtype=H.dtype, device=H.device).scatter_reduce_(
            self.dim, index_torch, alphas * H, reduce="sum", include_self=False
        )
agg = NewAgg(output_size=mp.output_dim)		
```

I'm note sure `attention_logits` is the best name for that variable. I'm not even sure that is a correct name, so open to suggestions.

Also note that I subtract `self.W(H).max()` before exponentiating. This doesn't change the output, but I think it helps with numerical stability? At least this is what [scatter_softmax](https://github.com/rusty1s/pytorch_scatter/blob/c095c62e4334fcd05e4ac3c4bb09d285960d6be6/torch_scatter/composite/softmax.py#L45) does. IMPORTANT: `torch_scatter` actually uses `scatter_max` to get the max from each index grouping (each molecule) in a batch while my version only get the single largest value across the whole `self.W(H)` vector. I couldn't find a way to get the max value for each grouping using only native pytorch. Using the max value of the whole vector may cause problems if one molecule's message passing output is much larger than the other molecules, so it would be good to find a more correct way to do this. 

Second IMPORTANT note, I found numerical instability when testing a tensor of all ones (`H = torch.ones_like(mp(bmg, V_d))`). The output should be a tensor of ones. Sometimes however some of the values are `1.0000001192092896` or `0.9999999403953552`. These are one bit above and below `1.0` (determined using [this method](https://stackoverflow.com/questions/16444726/binary-representation-of-float-in-python-bits-not-hex)). I also sometimes saw `0.9999992847442627`, `1.0000003576278687`, `1.0000004768371582`, `0.9999992251396179` and potentially others that are all off of `1.0` by a couple of bits. 

In the test below I simulate the output of message passing with a variable number of atoms, a batch size of 1 (so all molecule indices are 0, `batch = torch.zeros(n_atoms, dtype=torch.int64)` and an mp output of all ones (`H = torch.ones(n_atoms, mp_output_dim, dtype=torch.float32`). The dtypes are the same as what I get out of a `MolGraphDataLoader`. I compare three methods for the forward method of `AttentiveAggregation`. `forward_v2dev` is the current `torch_scatter` version. `forward_notscaled` is my version but without subtracting the max value before exponentiating. `forward_scaled` is the version I presented. The off-by-one-bit error doesn't happen for every random initialization of `W` (the attention layer that is just a `torch.nn.Linear(mp_output_dim, 1)`) so I loop through the test 1000 times to get a sense of how often it fails. My results are summarized below followed by the code.

| n_atoms | v2/dev errors | not scaled errors | scaled errors|
|---|---|---|---|
|3|0|326|0|
|4|0|0|0|
|5|0|193|0|
|6|0|442|0|
|7|0|466|0|
|8|0|447|0|
|9|0|375|0|
|10|1000|777|1000|
|11|1000|1000|1000|
|12|1000|1000|1000|
|13|0|432|0|
|14|1000|1000|1000|
|15|0|474|0|
|16|0|697|0|
|17|0|431|0|
|18|1000|1000|1000|
|19|1000|1000|1000|
|20|1000|832|1000|
|21|0|551|0|
|22|1000|585|1000|
|23|1000|1000|1000|
|24|1000|1000|1000|
|25|1000|1000|1000|
|26|1000|1000|1000|
|27|1000|1000|1000|
|28|1000|920|1000|
|29|1000|1000|1000|
|30|1000|1000|1000|
|31|1000|1000|1000|
|32|0|844|0|

I'll stop there, but I continued to see ups and downs in how often a couple bits are off. I'll note that most of the time `torch_scatter` and my scaled version have no failures while the unscaled version does, but there are also n_atoms values for which those two version fail everytime and the unscaled version only fails most of the time. 

Ultimately the fact that the `torch_scatter` version also has failures makes me think that we don't need to worry about this. And my proposed scaled version fails the same times that `torch_scatter` does so it could be a replacement. 

```
dim = 0
def forward_v2dev(H, batch, W):
    alphas = scatter_softmax(W(H), batch, dim)
    return scatter(alphas * H, batch, dim, reduce="sum")
def forward_notscaled(H, batch, W):
    dim_size = batch.max().int() + 1
    attention_logits = W(H).exp()
    Z = torch.zeros(dim_size, 1, dtype=H.dtype, device=H.device).scatter_reduce_(
        dim, batch.unsqueeze(1), attention_logits, reduce="sum", include_self=False
    )
    alphas = attention_logits / Z[batch]
    index_torch = batch.unsqueeze(1).repeat(1, H.shape[1])
    return torch.zeros(dim_size, H.shape[1], dtype=H.dtype, device=H.device).scatter_reduce_(
        0, index_torch, alphas * H, reduce="sum", include_self=False
    )
def forward_scaled(H, batch, W):
    dim_size = batch.max().int() + 1
    attention_logits = (W(H) - W(H).max()).exp()
    Z = torch.zeros(dim_size, 1, dtype=H.dtype, device=H.device).scatter_reduce_(
        dim, batch.unsqueeze(1), attention_logits, reduce="sum", include_self=False
    )
    alphas = attention_logits / Z[batch]
    index_torch = batch.unsqueeze(1).repeat(1, H.shape[1])
    return torch.zeros(dim_size, H.shape[1], dtype=H.dtype, device=H.device).scatter_reduce_(
        0, index_torch, alphas * H, reduce="sum", include_self=False
    )
```
```
n_atoms = 29

mp_output_dim = 1
fail_v2dev = 0
fail_notscaled = 0
fail_scaled = 0
n = 1000
for i in range(n):
    W = torch.nn.Linear(mp_output_dim, 1)
    batch = torch.zeros(n_atoms, dtype=torch.int64)
    H = torch.ones(n_atoms, mp_output_dim, dtype=torch.float32)
    v2dev = forward_v2dev(H, batch, W)
    notscaled = forward_notscaled(H, batch, W)
    scaled = forward_scaled(H, batch, W)
    if v2dev.item() != 1.0:
        fail_v2dev += 1
    if notscaled.item() != 1.0:
        fail_notscaled += 1
    if scaled.item() != 1.0:
        fail_scaled += 1
print(f"v2dev fails: {fail_v2dev}/{n}")
print(f"new fails: {fail_notscaled}/{n}")
print(f"fix fails: {fail_scaled}/{n}")
print(notscaled.item())
```